### DISCUSSION TITLE
[WASM + SQLite] Update on current status?

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#10640] [Q&A] [answered]
I recently have seen more activity again around SQLite support in UNO and liked to understand where we're at regarding the support for SQLite in UNO WASM in particular.
- I'm aware of the repo [here](https://github.com/unoplatform/Uno.sqlite-wasm)
- ... as well as the sample [here](https://github.com/unoplatform/Uno.Samples/tree/master/UI/SQLiteSample)

My questions rather are:
- where is the data persisted for UNO WASM + SQLite and what are the current limitations to that persistance?
- how do the recent [SQLite plans](https://sqlite.org/wasm/doc/tip/about.md) fit into here and what does that mean for UNO WASM?

Thanks in advance for shedding some light.

**Comment by jeromelaban** [comment, marked as answer]
> * where is the data persisted for UNO WASM + SQLite and what are the current limitations to that persistance?

The data is persisted [whereever you place it](https://github.com/unoplatform/Uno.Samples/blob/6e5faa904c7dc840caa8ecfaa4c4fd8f03a257ae/UI/SQLiteSample/SQLiteSample/SQLiteSample/SQLiteSample.Shared/MainPage.xaml.cs#L36). Persisted folders [are specified here](https://platform.uno/docs/articles/features/file-management.html?q=storage#webassembly-file-system).

> * how do the recent [SQLite plans](https://sqlite.org/wasm/doc/tip/about.md) fit into here and what does that mean for UNO WASM?

This particular project is mostly about a javascript binding, does not apply to binding to C#. We'll keep building a separate version until .NET supports multi-versioning properly, and WebAssembly properly supports dynamic module referencing.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban. That gives me something to work with

**Comment by DierkDroth** [comment]
@jeromelaban I did some reading + reviewed code/projects etc.

Here is what I understand:
- all of the folder persistence options mentioned in [your link](https://github.com/unoplatform/Uno.Samples/blob/6e5faa904c7dc840caa8ecfaa4c4fd8f03a257ae/UI/SQLiteSample/SQLiteSample/SQLiteSample/SQLiteSample.Shared/MainPage.xaml.cs#L36) above would go to the respective 'folder' in the browser's IndexedDB
- none of them would actually go to the file system of the local device

This would come with the well-known limitations like:
- it's hard and close to impossible to implement a working backup strategy for line-of-business data stored in the DB (I'm aware that IndexedDB e.g. for Chrome has a representation in the file system)
- the risk of inadvertent deletion of the whole DB by a simple "clear browser cache" triggered by the user is unbearable
- browser may ask for confirmation if space limits are exceeded or even block further access altogether

I also understand that the 'native webassembly support' by SQLite does (theoretically) support access [to the local file system](https://sqlite.org/wasm/doc/trunk/persistence.md#coop-coep) (= OPFS). Hence, the limitations above would not apply.

I know this is not exactly UNO's territory. However, UNO provided SQLite access options and I wanted to make sure I'm not missing a potential option.

Could you please confirm/deny my findings?

**Reply by jeromelaban** [reply]
> the risk of inadvertent deletion of the whole DB by a simple "clear browser cache" triggered by the user is unbearable

Any kind of local storage is up to the user to manage. If you don't want the user to manage storage, you'll need to implement a server-side synchronization mechanism.

> I also understand that the 'native webassembly support' by SQLite does (theoretically) support access [to the local file system](https://sqlite.org/wasm/doc/trunk/persistence.md#coop-coep) (= OPFS). Hence, the limitations above would not apply.

Emscripten has some [partial support for OPFS](https://emscripten.org/docs/api_reference/Filesystem-API.html#new-file-system-wasmfs), but the version used by the .NET runtime has a version that may not be fully featured as a result. The WasmFS feature is (even as of latest emscripten) considered in progress. 

Also note that OPFS will not solve any of the possible clearing of the storage (see [this document](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API#origin_private_file_system)).

> I know this is not exactly UNO's territory. However, UNO provided SQLite access options and I wanted to make sure I'm not missing a potential option.

Uno provides a build of SQLite that uses available APIs from the underlying tooling, making the security constraints of the Web are applying fully.

**Reply by DierkDroth** [reply]
Thanks again for your feedback @jeromelaban. In particular [your link above](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API#origin_private_file_system) sheds some light on OPFS.

Browsing through the docs I wonder what is the point of OPFS if the same limitations as with IndexedDB still apply like
- quotas (potentially)
- risk of inadvertent deletion of data
- no clear mapping to file(s) on the local HD -> still the question remains how to backup/restore the DB (if needed)

Having OPFS just for improved improved (write) performance? Then why no making IndexedDB faster? ... oh well

> you'll need to implement a server-side synchronization mechanism.

FYI, there already is a 'cloud based' storage with my app. I'm looking for a 'local only' alternate storage option - similar to the desktop edition of my app which has its SQLite based local DB as well
